src_c/draw.c

```diff
@@ -1557,88 +1554,111 @@ compare_int(const void *a, const void *b)
+
+        /* Determine X bounds */
+        minx = point_x[0];
+        maxx = point_x[0];
+        for (i = 1; (i < num_points); i++) {
+            minx = MIN(minx, point_x[i]);
+            maxx = MAX(maxx, point_x[i]);
+        }
+        drawhorzlineclip(dst, color, minx, miny, maxx);
+        PyMem_Free(x_intersect);
+        return;
+    }
+
+    /* Draw, scanning y
+     * ----------------
+     * The algorithm uses a horizontal line (y) that moves from top to the
+     * bottom of the polygon:
+     *
+     * 1. search intersections with the border lines
+     * 2. sort intersections (x_intersect)
+     * 3. each two x-coordinates in x_intersect are then inside the polygon
+     *    (drawhorzlineclip for a pair of two such points)
+     */
+    // n_intersections : number of intersections with the polygon
+    int n_intersections;
     for (y = miny; (y <= maxy); y++) {
-        ints = 0;
-        for (i = 0; (i < n); i++) {
-            if (!i) {
-                ind1 = n - 1;
-                ind2 = 0;
-            }
-            else {
-                ind1 = i - 1;
-                ind2 = i;
-            }
-            y1 = vy[ind1];
-            y2 = vy[ind2];
+        n_intersections = 0;
+        for (i = 0; (i < num_points); i++) {
+            i_previous = ((i) ? (i - 1) : (num_points - 1));
+
+            y1 = point_y[i_previous];
+            y2 = point_y[i];
             if (y1 < y2) {
-                x1 = vx[ind1];
-                x2 = vx[ind2];
+                x1 = point_x[i_previous];
+                x2 = point_x[i];
             }
             else if (y1 > y2) {
-                y2 = vy[ind1];
-                y1 = vy[ind2];
-                x2 = vx[ind1];
-                x1 = vx[ind2];
-            }
-            else if (miny == maxy) {
-                /* Special case: polygon only 1 pixel high. */
-                int j;
-                int minx, maxx;
-
-                /* Determine X bounds */
-                minx = vx[0];
-                maxx = vx[0];
-                for (j = 1; (j < n); j++) {
-                    minx = MIN(minx, vx[j]);
-                    maxx = MAX(maxx, vx[j]);
-                }
-
-                /* Just a line from minimum to maximum X */
-                polyints[ints++] = minx;
-                polyints[ints++] = maxx;
-                break;
+                y2 = point_y[i_previous];
+                y1 = point_y[i];
+                x2 = point_x[i_previous];
+                x1 = point_x[i];
             }
-            else {
+            else { // y1 == y2 : has to be handled as special case (below)
                 continue;
             }
-            if ((y >= y1) && (y < y2)) {
-                polyints[ints++] = (y - y1) * (x2 - x1) / (y2 - y1) + x1;
-            }
-            else if ((y == maxy) && (y > y1) && (y <= y2)) {
-                polyints[ints++] = (y - y1) * (x2 - x1) / (y2 - y1) + x1;
+            if ( ((y >= y1) && (y < y2)) ||
```

Comment:
One way to make expressions a bit smaller is to give names to their components.
```c
bool y_crosses_edge = ((y >= y1) && (y < y2))
bool on_lowest_line = ((y == maxy) && (y2 == maxy))
if (y_crosses_edge || on_lowest_line) {
```
However, I wonder if C compilers are smart enough to opitimze this well... ?